File: leetcode_coding_challang/two_pointer_approach_medium/777_swap_adjacent_in_lr_string/solution.js
/**
 * @param {string} start
 * @param {string} result
 * @return {boolean}
 */
var canTransform = function (start, result) {
  const n = start.length;

  // Two pointers: i for start, j for result
  let i = 0,
    j = 0;

  while (i < n || j < n) {
    // Skip all 'X' in start string
    while (i < n && start[i] === "X") {
      i++;
    }

    // Skip all 'X' in result string
    while (j < n && result[j] === "X") {
      j++;
    }

    // If both pointers reached the end, transformation is valid
    if (i === n && j === n) {
      return true;
    }

    // If only one pointer reached the end, invalid
    if (i === n || j === n) {
      return false;
    }

    // Characters at current positions must match
    if (start[i] !== result[j]) {
      return false;
    }

    // Check movement constraints:
    // 'L' can only move left (result position <= start position)
    if (start[i] === "L" && i < j) {
      return false;
    }

    // 'R' can only move right (result position >= start position)
    if (start[i] === "R" && i > j) {
      return false;
    }

    // Move both pointers forward
    i++;
    j++;
  }

  return true;
};

// Test cases
console.log(canTransform("RXXLRXRXL", "XRLXXRRLX")); // true
console.log(canTransform("X", "L")); // false
console.log(canTransform("XXXXXLXXXX", "LXXXXXXXXX")); // true

File: leetcode_coding_challang/two_pointer_approach_medium/777_swap_adjacent_in_lr_string/solution.java
class Solution {
    /**
     * Determines if start string can be transformed to result string
     * using allowed swap operations: XL->LX and RX->XR
     * 
     * @param start The starting string
     * @param result The target result string
     * @return true if transformation is possible, false otherwise
     */
    public boolean canTransform(String start, String result) {
        int n = start.length();
        
        // Two pointers: i for start, j for result
        int i = 0, j = 0;
        
        while (i < n || j < n) {
            // Skip all 'X' characters in start string
            while (i < n && start.charAt(i) == 'X') {
                i++;
            }
            
            // Skip all 'X' characters in result string
            while (j < n && result.charAt(j) == 'X') {
                j++;
            }
            
            // Both pointers reached the end - valid transformation
            if (i == n && j == n) {
                return true;
            }
            
            // Only one pointer reached the end - invalid
            if (i == n || j == n) {
                return false;
            }
            
            // The non-X characters must match at corresponding positions
            if (start.charAt(i) != result.charAt(j)) {
                return false;
            }
            
            // Movement constraint for 'L':
            // Can only move left, so position in result must be <= position in start
            if (start.charAt(i) == 'L' && i < j) {
                return false;
            }
            
            // Movement constraint for 'R':
            // Can only move right, so position in result must be >= position in start
            if (start.charAt(i) == 'R' && i > j) {
                return false;
            }
            
            // Move both pointers to next positions
            i++;
            j++;
        }
        
        return true;
    }
    
    // Test the solution
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        System.out.println(solution.canTransform("RXXLRXRXL", "XRLXXRRLX")); // true
        System.out.println(solution.canTransform("X", "L")); // false
        System.out.println(solution.canTransform("XXXXXLXXXX", "LXXXXXXXXX")); // true
    }
}

